src/main/java/com/google/firebase/internal/RetryInitializer.java

```diff
@@ -0,0 +1,106 @@
+ * to the {@code HttpUnsuccessfulResponseHandler} that was originally set on the request. If
+ * the request does not get retried at that level, {@link RetryUnsuccessfulResponseHandler} is used
+ * to schedule additional retries.
+ */
+final class RetryInitializer implements HttpRequestInitializer {
+
+  private final RetryConfig retryConfig;
+
+  RetryInitializer(RetryConfig retryConfig) {
+    this.retryConfig = checkNotNull(retryConfig);
+  }
+
+  @Override
+  public void initialize(HttpRequest request) {
+    request.setNumberOfRetries(retryConfig.getMaxRetries());
+    request.setUnsuccessfulResponseHandler(newUnsuccessfulResponseHandler(request));
+    request.setIOExceptionHandler(newIOExceptionHandler());
+  }
+
+  private HttpUnsuccessfulResponseHandler newUnsuccessfulResponseHandler(HttpRequest request) {
+    RetryUnsuccessfulResponseHandler retryHandler = new RetryUnsuccessfulResponseHandler(
+        retryConfig);
+    return new RetryHandlerDecorator(retryHandler, request);
+  }
+
+  private HttpIOExceptionHandler newIOExceptionHandler() {
+    return new HttpBackOffIOExceptionHandler(retryConfig.newBackOff())
+        .setSleeper(retryConfig.getSleeper());
+  }
+
+  /**
+   * Makes sure that any error handlers already set on the request are executed before the retry
+   * handler is called. This is needed since some initializers (e.g. HttpCredentialsAdapter)
+   * register their own error handlers.
+   */
+  private static class RetryHandlerDecorator implements HttpUnsuccessfulResponseHandler {
+
+    private final RetryUnsuccessfulResponseHandler retryHandler;
+    private final HttpUnsuccessfulResponseHandler preRetryHandler;
+
+    private RetryHandlerDecorator(
+        RetryUnsuccessfulResponseHandler retryHandler, HttpRequest request) {
+      this.retryHandler = checkNotNull(retryHandler);
+      HttpUnsuccessfulResponseHandler preRetryHandler = request.getUnsuccessfulResponseHandler();
+      if (preRetryHandler == null) {
+        preRetryHandler = new HttpUnsuccessfulResponseHandler() {
+          @Override
+          public boolean handleResponse(
+              HttpRequest request, HttpResponse response, boolean supportsRetry) {
+            return false;
+          }
+        };
+      }
+      this.preRetryHandler = preRetryHandler;
+    }
+
+    @Override
+    public boolean handleResponse(
+        HttpRequest request,
+        HttpResponse response,
+        boolean supportsRetry) throws IOException {
+      try {
+        boolean retry = preRetryHandler.handleResponse(request, response, supportsRetry);
+        if (!retry) {
+          retry = retryHandler.handleResponse(request, response, supportsRetry);
+        }
+        return retry;
+      } finally {
+        // Pre-retry handler may have reset the unsuccessful response handler on the
+        // request. This changes it back.
+        request.setUnsuccessfulResponseHandler(this);
```

Comment:
Do we need to update `preRetryHandler` and replace it with the current retry handler?

Reply:
No it should be initialized only once per request.